### PR TITLE
Firewall Profile Implementation

### PR DESCRIPTION
**Pull request by adplotzk**
Firewall Profile Description of Changed:
1. Adding firewall.json in profiles
2. Adding firewall_rule.json in objects which extends the rule object for the firewall profile usecase. 
3. Updating http.json in events/network to include the application of the firewall profile in to accommodate WAF events. 
4. Updating disctionary.json to include new attributes from firewall.json profile and firewall_rule.json object, as well as adding additional disposition_id options to reflect possible firewall actions. 

The following is an example AWS WAF log conforming to the firewall profile as applies to 4002.

    {
        "activity_id": 1,
        "activity_name": "Get",
        "category_name": "Network Activity",
        "category_uid": 4,
        "class_name": "HTTP Activity",
        "class_uid": 4002,
        "type_name": "HTTP Activity: Get",
        "type_uid": 400503,
        "time": 1576280412771,
        "disposition": "Block",
        "disposition_id": "2",
        "time": 1576280412771,
        "metadata": {
            "product": {
	            "name": "AWS WAF",
	            "vendor_name": "AWS"
            },
            "feature": {
	            "uid": "arn:aws:wafv2:ap-southeast-2:111122223333:regional/webacl/STMTest/1EXAMPLE-2ARN-3ARN-4ARN-123456EXAMPLE"
            },
            "log_version": "1",
            "profiles": [firewall],
            "version": "1.0.0-rc.3",
            "labels": [
	            {
		            "name": "value"
	            }
            ]
        },
        "src_endpoint": {
            "ip": "1.1.1.1",
            "location": {
	            "country": "AU"
            }
        },
        "http_request": {
            "args": "",
            "version": "HTTP/1.1",
            "method": "GET",
            "uid": "rid",
            "user_agent": "curl/7.61.1",
            "url": {
	            "hostname": "localhost:1989",
	            "path": "/myUri",
	            "url_string": "localhost:1989/myUri"
            },
            "headers": [
	            {
		            "name": "Accept",
		            "value": "*/*"
	            },
	            {
		            "name": "x-stm-test",
		            "value": "10 AND 1=1"
	            }
            ]
        },
        "rule": {
            "uid": "STMTest_SQLi_XSS",
            "type": "REGULAR",
            "condition": "SQL_INJECTION",
            "sensitivity": "HIGH",
            "location": "HEADER",
            "matched_data": [
	            "10",
	            "AND",
	            "1"
            ]
        },
        "unmapped": {
            "httpSourceName": "APIGW",
            "httpSourcId": "-"
        }
    }

### CODE REVIEW

File: objects/firewall_rule.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  "attributes": {
+    "condition": {
+      "description": "The rule trigger condition for the rule. For example: SQL_INJECTION.",
+      "requirement": "optional"
+    },
+    "sensitivity": {
+      "description": "The location in the request where the rule matched. For example: HIGH.",
+      "requirement": "optional"
+    },	
+    "rule_location": {
```
Comment (rroupski) [inline objects/firewall_rule.json:15]: Based on the description, it sounds like the name should be `match_location`?

File: objects/firewall_rule.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  "name": "firewall_rule",
+  "description": "The firewall rule object associated with a policy or event.",
+  "extends": "rule",
+  "attributes": {
+    "condition": {
+      "description": "The rule trigger condition for the rule. For example: SQL_INJECTION.",
+      "requirement": "optional"
+    },
+    "sensitivity": {
+      "description": "The location in the request where the rule matched. For example: HIGH.",
```
Comment (floydtree) [inline objects/firewall_rule.json:12]: Same as earlier, appears to be a duplicate of rule_location description. Unclear what sensitivity means.

File: dictionary.json [hunk 3]
```diff
@@ -3227,6 +3233,32 @@
...
+    "match_details": {
+      "caption": "Match Details",
+      "description": "The data in a request that rule matched. For example: '[\"10\",\"and\",\"1\"]'.",
+      "is_array": true,
+      "type": "string_t"
+    },
+    "rate_limit": {
+      "caption": "Rate Limit",
+      "description": "The rate limit for a rate-based rule.",
+      "type": "string_t"
```
Comment (rroupski) [inline dictionary.json:3261]: Is the value a number? Should the type be an integer?

File: objects/firewall_rule.json (hunk 2 above)
Comment (rroupski) [inline objects/firewall_rule.json:12]: Wrong description?

File: dictionary.json
Comment (floydtree) [inline dictionary.json:1]: You would need to define firewall_rule attribute in the dictionary as well. Refer to the contribution guide.

I recommend running a local instance of the server to see if the run is error/warning free.

```
Warning: 18:18:31.719 [warning] 'HTTP Activity' uses undefined attribute: firewall_rule: %{_source: :http_activity, group: "primary", profile: "firewall", requirement: "required"}
Warning: 18:18:31.722 [warning] profile firewall uses firewall_rule that is not defined in the dictionary
```

File: objects/firewall_rule.json [hunk 4]
```diff
@@ -0,0 +1,40 @@
+{
```
Comment (rroupski) [inline objects/firewall_rule.json:1]: Remove the attribute descriptions that are the same as the ones defined in the dictionary.

**Review by floydtree — CHANGES_REQUESTED** [review]
Specific comments inline.

General - as I mentioned in the closed PR - 

> I would also recommend to edit the PR description and highlight what exactly has been changed, added, removed etc. This would help in generating clearer & cleaner release notes.

File: dictionary.json [hunk 5]
```diff
@@ -3227,6 +3233,32 @@
...
       "type": "string_t"
+    },
+    "condition": {
+      "caption": "Condition",
+      "description": "The rule trigger condition for the rule. For example: SQL_INJECTION.",
+      "type": "string_t"
+    },
+    "sensitivity": {
+      "caption": "Sensitivity",
+      "description": "The location in the request where the rule matched. For example: HIGH.",
```
Comment (rroupski) [inline dictionary.json:3244]: Copy and paste description?

File: dictionary.json [hunk 6]
```diff
@@ -3227,6 +3232,32 @@
...
+      "type": "string_t"
+    },
+    "sensitivity": {
+      "caption": "Sensitivity",
+      "description": "The sensitivity of the firewall rule in the matched event. For example: HIGH.",
+      "type": "string_t"
+    },
+    "match_location": {
+      "caption": "Match Location",
+      "description": "The location of the in the matched data in the event as it pertains to the firewall rule. For example: HEADER.",
```
Comment (floydtree) [inline dictionary.json:3248]: Mistake in the description? Not clear what it means

File: objects/firewall_rule.json [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  "extends": "rule",
+  "attributes": {
+    "condition": {
+      "requirement": "optional"
+    },
+    "sensitivity": {
+      "requirement": "optional"
+    },	
+    "match_location": {
+      "description": "The location of the in the matched data in the event as it pertains to the firewall rule. For example: HEADER.",
```
Comment (floydtree) [inline objects/firewall_rule.json:14]: no need for description here, if it's the same as the one in dictionary

**Review by floydtree — CHANGES_REQUESTED** [review]

**Review by rroupski — DISMISSED** [review]

**Review by mr-splunk — DISMISSED** [review]

**Review by floydtree — APPROVED** [review]

**Review by rroupski — APPROVED** [review]

**Review by mr-splunk — APPROVED** [review]

**Pull request merged**